src/core/symbology/qgssymbollayerutils.cpp

```diff
@@ -4730,3 +4730,44 @@ QgsSymbol *QgsSymbolLayerUtils::restrictedSizeSymbol( const QgsSymbol *s, double
   }
   return 0;
 }
+
+QgsStringMap QgsSymbolLayerUtils::evaluatePropertiesMap( const QMap<QString, QgsProperty> &propertiesMap, const QgsExpressionContext &context )
+{
+  QgsStringMap properties;
+  QMap<QString, QgsProperty>::const_iterator paramIt = propertiesMap.constBegin();
+  for ( ; paramIt != propertiesMap.constEnd(); ++paramIt )
+  {
+    properties.insert( paramIt.key(), paramIt.value().valueAsString( context ) );
+  }
+  return properties;
+}
+
+QMap<QString, QString> QgsSymbolLayerUtils::serializeProperties( const QMap<QString, QgsProperty> &propertiesMap )
+{
+  QMap<QString, QString> serialized;
+  QMap<QString, QgsProperty>::const_iterator paramIt = propertiesMap.constBegin();
+  int i = 0;
+  for ( ; paramIt != propertiesMap.constEnd(); ++paramIt )
+  {
+    serialized[QString( QStringLiteral( "properties_%1_name" ) ).arg( i )] = paramIt.key();
+    serialized[QString( QStringLiteral( "properties_%1_value" ) ).arg( i )] = paramIt.value().expressionString();
```

Comment:
I don't like the forced saving of expression string here -- from the docs "If the property is not a ExpressionBasedProperty this will return an empty string.". So this is a potential source of future bugs, as the api which allows setting a QMap< QString, QgsProperty > on a symbol naturally implies that any type of property is acceptable...

Better to use QgsProperty.toVariant and fromVariant here so that there's no loss of QgsProperty properties.

Reply:
But that's a bit hard to fit this in current API:

`typedef QgsSymbolLayer *( *QgsSymbolLayerCreateFunc )( const QgsStringMap & ) SIP_SKIP;`

The create method expects a QgsStringMap. How would you do this?

Reply:
The only reasonable approach I see would to change in `QgsSymbolLayer`:
`virtual QgsStringMap properties() const = 0;`
to
`virtual QVariantMap properties() const = 0;`

(and probably keep the old signature for Python bindings)

that means, that `QgsSymbolLayerUtils::saveSymbol`, we would use `QgsXmlUtils::writeVariant` rather than `saveProperties`.

Does this make sense to you?

Reply:
I'd love to see the change from string map to variant map -- it frustrates me that trying to create a symbol from

    sym = QgsFillSymbol.createSimple( { 'color' :...., 'stroke_width': 1.2 } )

fails because you have to quote the numeric values. (You'd also need to update the .create methods to match). Not sure how big a job that is though...

An ugly alternative is to convert to xml and then dump the xml into the string value. I won't block you if you want to do this -- I've had to resort to this in the past too...

Reply:
that was quite some work.... #40716 done!